### PR TITLE
[AIRAVATA-2659] GroupResourceProfile for modeling resource sharing with specific groups of users

### PR DESCRIPTION
**Pull request by sachinkariyattin**
The GroupResourceProfile implementation has been added to the app_catalog in the registry module and the following changes have been implemented.

- Created the following thrift models for the GroupResourceProfile
      `GroupResourceProfile`
      `GroupComputeResourcePreference`
      `ComputeResourcePolicy`
      `BatchQueueResourcePolicy`
      `GroupAccountSSHProvisionerConfig `

- Created new tables in the app_catalog schema. (GROUP_RESOURCE_PROFILE, BATCH_QUEUE_RESOURCE_POLICY, COMPUTE_RESOURCE_POLICY, GROUP_COMPUTE_RESOURCE_PREFERENCE, GRP_SSH_ACC_PROV_CONFIG)
- Implemented GroupResourceProfileRepository.
- Added GroupResourceProfileRepositoryTest.
- Added the following methods to the Airavata API related to the GroupResourceProfile
    ```
     createGroupResourceProfile()
     updateGroupResourceProfile()
     getGroupResourceProfile()
     removeGroupResourceProfile()
     getGroupResourceList()
     removeGroupComputePrefs()
     removeGroupComputeResourcePolicy()
     removeGroupBatchQueueResourcePolicy()
     getGroupComputeResourcePreference()
     getGroupComputeResourcePolicy()
     getBatchQueueResourcePolicy()
     getGroupComputeResourcePrefList()
     getGroupBatchQueueResourcePolicyList()
     getGroupComputeResourcePolicyList()
    ```

### CODE REVIEW

**Comment by machristie** [comment]
@eldhomathulla here's Sachin's pull request. See especially his changes to Thrift files.

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 1]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
   PRIMARY KEY (RESOURCE_ID,USER_ID,GATEWAY_ID)
 ) ENGINE=InnoDB DEFAULT CHARSET=latin1;
 
+CREATE TABLE GROUP_RESOURCE_PROFILE (
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Let's make the PK GROUP_RESOURCE_PROFILE_ID, and then a unique constraint on (GATEWAY_ID, GROUP_RESOURCE_PROFILE_NAME).

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 2]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
...
+  GATEWAY_ID varchar(255) NOT NULL,
+  GROUP_RESOURCE_PROFILE_ID varchar(255) NOT NULL,
+  GROUP_RESOURCE_PROFILE_NAME varchar(255) DEFAULT NULL,
+  CREATION_TIME BIGINT NOT NULL,
+  UPDATE_TIME BIGINT NOT NULL,
+  PRIMARY KEY (GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID)
+)ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+CREATE TABLE BATCH_QUEUE_RESOURCE_POLICY (
+  RESOURCE_POLICY_ID varchar(255) NOT NULL,
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Can you make the PK just RESOURCE_POLICY_ID? And can you add a FK to COMPUTE_RESOURCE?

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 3]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
...
+  MAX_ALLOWED_NODES INTEGER,
+  MAX_ALLOWED_CORES INTEGER,
+  MAX_ALLOWED_WALLTIME INTEGER,
+  PRIMARY KEY (RESOURCE_POLICY_ID,COMPUTE_RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID,GATEWAY_ID,QUEUE_NAME),
+  FOREIGN KEY (COMPUTE_RESOURCE_ID,QUEUE_NAME) REFERENCES BATCH_QUEUE(COMPUTE_RESOURCE_ID,QUEUE_NAME) ON DELETE CASCADE,
+  FOREIGN KEY (GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID) REFERENCES GROUP_RESOURCE_PROFILE(GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID) ON DELETE CASCADE
+)ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+CREATE TABLE COMPUTE_RESOURCE_POLICY (
+    RESOURCE_POLICY_ID varchar(255) NOT NULL,
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Can you make the PK just the RESOURCE_POLICY_ID column?

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 4]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
...
+    COMPUTE_RESOURCE_ID varchar(255) NOT NULL,
+    GROUP_RESOURCE_PROFILE_ID varchar(255) NOT NULL,
+    GATEWAY_ID varchar(255) NOT NULL,
+    PRIMARY KEY (GATEWAY_ID,RESOURCE_POLICY_ID,COMPUTE_RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID),
+    FOREIGN KEY (COMPUTE_RESOURCE_ID) REFERENCES COMPUTE_RESOURCE(RESOURCE_ID) ON DELETE CASCADE,
+    FOREIGN KEY (GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID) REFERENCES GROUP_RESOURCE_PROFILE(GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID) ON DELETE CASCADE
+)ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+CREATE TABLE COMPUTE_RESOURCE_POLICY_QUEUES (
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Let's make a FK to BATCH_QUEUE here. And the PK can be (RESOURCE_POLICY_ID, QUEUE_NAME).

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 5]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
...
+    GROUP_RESOURCE_PROFILE_ID varchar(255) NOT NULL,
+    GATEWAY_ID varchar(255) NOT NULL,
+    QUEUE_NAME varchar(255),
+    PRIMARY KEY (GATEWAY_ID,RESOURCE_POLICY_ID,COMPUTE_RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID),
+    FOREIGN KEY (GATEWAY_ID,RESOURCE_POLICY_ID,COMPUTE_RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID) REFERENCES COMPUTE_RESOURCE_POLICY(GATEWAY_ID,RESOURCE_POLICY_ID,COMPUTE_RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID) ON DELETE CASCADE
+)ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+CREATE TABLE GROUP_COMPUTE_RESOURCE_PREFERENCE
+(
+        RESOURCE_ID VARCHAR(255) NOT NULL,
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Primary key should just be (RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID).

File: modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql [hunk 6]
```diff
@@ -568,6 +568,85 @@ CREATE TABLE USER_COMPUTE_RESOURCE_PREFERENCE (
...
+        RESERVATION_START_TIME timestamp,
+        RESERVATION_END_TIME timestamp,
+        SSH_ACCOUNT_PROVISIONER VARCHAR(255),
+        SSH_ACCOUNT_PROVISIONER_ADDITIONAL_INFO VARCHAR(1000),
+        PRIMARY KEY(RESOURCE_ID,GROUP_RESOURCE_PROFILE_ID,GATEWAY_ID),
+        FOREIGN KEY (RESOURCE_ID) REFERENCES COMPUTE_RESOURCE(RESOURCE_ID) ON DELETE CASCADE,
+        FOREIGN KEY (GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID) REFERENCES GROUP_RESOURCE_PROFILE(GATEWAY_ID,GROUP_RESOURCE_PROFILE_ID)
+)ENGINE=InnoDB DEFAULT CHARSET=latin1;
+
+CREATE TABLE GRP_SSH_ACC_PROV_CONFIG
```
Comment (machristie) [inline modules/registry-refactoring/src/main/resources/appcatalog-mysql.sql]: Let's make the PK here (GROUP_RESOURCE_PROFILE_ID , RESOURCE_ID, CONFIG_NAME)

**Comment by sachinkariyattin** [comment]
@machristie Thanks for the review. 
I have implemented the suggested changes.

**Comment by machristie** [comment]
Nice job Sachin!

**Pull request merged**